toolchain/parse/typed_nodes.h

```diff
@@ -1032,6 +1034,51 @@ struct DesignatorExpr {
   NodeIdOneOf<IdentifierName, SelfTypeName> name;
 };
 
+// FIXME: RequirementRewrite?
+struct RequirementAssign {
+  static constexpr auto Kind = NodeKind::RequirementAssign.Define(
+      {.category = NodeCategory::Requirement, .child_count = 2});
+  AnyExprId lhs;
```

Comment:
Should we syntactically require the LHS to be a `DesignatorExpr`? (I'm fine for that to be future work rather than something done in this PR.)

Reply:
I was thinking that should be done in check. I think the alternative would require a lot of special handling in parse. I assume the decision comes down to error handling, which I assumed would be easier in check.

Reply:
If this is a syntactic restriction (that is, the syntax must be exactly `where .Foo = <expr>`, then I think that belongs in parse. If it's a semantic requirement, and we would allow (for example) `where (.Self.(SelfType.Foo)) = <expr>`, or any other expression that symbolically evaluates to the name of an associated constant in `.Self`, then I think that belongs in check. (I would expect that the only way for check to reliably distinguish these two examples would be by looking at the parse tree nodes after the fact, and that's something that I think check should ideally not be doing.)

As it stands, I think the rule is syntactic: [the design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#rewrite-constraints) says "In a rewrite constraint, the left-hand operand of `=` must be a `.` followed by the name of an associated constant."

Reply:
Added TODO